Fixes #3551 - Show unavailable Puppet Classes in Host Group edit

bastilian opened the pull request:
Puppet Classes can be assigned to a Host Group even if they are
not in the set environment. They will persist through out changes
of the environment, but were previously hidden from the UI.

Unavailable Puppet Classes are now shown in the UI as unavailable
and a warning is triggered to inform the user about them.

**Host Group edit form / Puppet Classes with changed environment:**
![screen shot 2016-11-24 at 16 08 21](https://cloud.githubusercontent.com/assets/7757/20603126/66e1e54a-b260-11e6-9618-122f1d9f6c45.png)

**Warning:**
![screen shot 2016-11-24 at 16 08 10](https://cloud.githubusercontent.com/assets/7757/20603125/66e18280-b260-11e6-8c15-d1239eb8e0d5.png)

dLobatog commented:
@theforeman-bot ok to test

dLobatog commented:
@domcleal @ohadlevy Could you add @bastilian to @theforeman org please?

dLobatog commented on app/models/concerns/host_common.rb:
```diff
@@ -201,16 +201,25 @@ def classes_in_groups
+  #  * +with_environments+: {Boolean} if possible environments should be included
+  #  * +in_environments+: {Boolean} to consider only classes in environments
+  #
+  def individual_puppetclasses(with_environments: false, in_environment: true)
```
I'm not sure I understand why this method has arguments if it's only called once (in the view), where 'with_environments' is true and 'in_environment' is false. I'd suggest to either:

* remove the arguments completely (adds complexity to the method that I don't think we're using anywhere)
* make the arguments default to what we're using

bastilian replied:
Ah, you are right! I overlooked that it is just used once in the whole app.

I'll rewrite it to not need arguments.

dLobatog commented on test/controllers/hostgroups_controller_test.rb:
```diff
@@ -141,16 +141,37 @@ def setup_user(operation, type = 'hostgroups')
+      assert_equal 0, @hostgroup.puppetclasses.length
+
+      post :environment_selected, @params, set_session_user
+      assert_equal(1, (assigns(:hostgroup).puppetclasses.length))
```
Nitpick, you probably want to check `assert_include @puppetclass, assigns(:hostgroup).puppetclasses` to ensure it's the right class

dLobatog replied:
fixed

dLobatog reviewed (changes requested):
@bastilian Thank you, this works pretty well from what I've seen. 

Once you save the host group, when you edit it again, the Puppet classes are colored in red, but it doesn't tell you why. You only see the 'unavailable' warning message when you change the environment, after that it's hard to know.

Can you add an `alert(:class => 'alert-warning'` to display this information? (or any equivalent way to show the classes are indeed unavailable) The rest of the PR is pretty much ready!

dLobatog commented on app/assets/javascripts/application.js:
```diff
@@ -351,6 +351,12 @@ $(function() {
 
+function check_for_unavailable_puppetclasses() {
+  if ($('#puppet_klasses #selected_classes .unavailable').size() > 0) {
+    notify('<span>' + __('Some Puppet Classes are unavailable in the selected environment') + '</span>', 'warning')
```
I think it's 'Puppet class' (lowercase class) - we're certainly not being very consistent about this around the app but a quick search revealed it's only 'Puppet Class' on a couple of titles, not texts

bastilian replied:
I've taken the tab title, that capitalises both, considering that is what the user need to look for when the alert pops up.

bastilian replied:
BUT. I do think that "Puppet class" would be the correct way to write it, as "Puppet" is a "brand" and "class" is just another word.

ohadlevy commented:
@bastilian why do you show the error as a toast notification? shouldn't it be within (or next) to the puppet class section on the form?
also, does red (in your image) means you can't remove it, or disabled? I was unsure what it means from looking at your screenshot. if its unavialavble, shouldn't it be disabled? /cc @rohoover

Rohoover commented:
@ohadlevy 
I interpreted the red text as being part of the form error. Perhaps I'm mistaken.

bastilian commented:
@ohadlevy  & @Rohoover yeah, it is not obvious why they are red, @dLobatog pointed that out as well.

I'm gonna add a note on the Puppet classes tab elaborating on why they are red and that they are going to be wiped if submitted.

ohadlevy commented:
can you explain it here as well? maybe @Rohoover could suggest an
alternative approach? thanks!

bastilian commented:
The case here is that the user has changed the environment of the host group. The environment that has been selected environment does not include the puppet classes, that have been previously configured for the host group with an environment that had these classes.

Marking these red may not be appropriate, now that I think about it, as the are actually not wiped, but are just "ineffective" in the set environment not including them. 

In case the environment changes to one that includes them they will be considered again.

Let me know if you have more questions to find a good way to communicate this in the UI, @Rohoover.

Rohoover commented:
@bastilian 

A few thoughts:

Placing an inline notification at the top (yellow) would help give the user more insight. (http://www.patternfly.org/pattern-library/communication/inline-notifications/#/api)
After you explanation - I agree, red may not be accurate. Perhaps grey them out?
Lastly, in your example no included classes were applicable. What remedies does the user have other than to switch environments?

bastilian commented:
@Rohoover Switching the environment is the only way to have them active again.

Another action the user could take is to remove the unavailable classes to avoid mistakenly activate them by updating the environment, later on. A link offering that in the message explaining the state would make sense I think.

dLobatog commented on app/assets/stylesheets/puppetclasses.scss:
```diff
@@ -32,8 +32,12 @@
 }
+
+.selected_puppetclass.unavailable a {
+  color: #cc0000
```
http://www.patternfly.org/styles/color-palette/#_ I believe you can use one of these

dLobatog commented:
Set to WoC @bastilian if you can find an way to show the 'resolution' (maybe just link to the primary tab and highlight the Environment field) with js is good - @Rohoover is that what you're looking for?

Rohoover commented:
@bastilian @dLobatog 

I think offering the user the feedback in the notification will go a long way. If it's possible to link to the tab within that notification then that could also be helpful.

ohadlevy commented on app/assets/javascripts/application.js:
```diff
@@ -338,6 +338,26 @@ $(function() {
 });
 
+function check_for_unavailable_puppetclasses() {
+  var unavailable_classes = $('#puppet_klasses #selected_classes .unavailable');
```
tbh: I think this needs to be rewritten in webpack, this way you can add tests to the function...

bastilian replied:
Yes, I am also in favour of rewriting it under webpack, but wouldn't it be better to to have a component for the edit form in general first and then move over components like in this case for editing Puppet classes?

To have this tested now we can use an integration test, which can be removed (or not) once it is rewritten with tests under webpack.

dLobatog replied:
I have to agree with @ohadlevy here, the function is relatively complicated so if we can get unit tests for them that's great. It doesn't have to be a React component, just moving the function there would make sense IMO so you can use Jest to test it. Like this one: https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/foreman_users.js

ohadlevy commented on app/assets/stylesheets/puppetclasses.scss:
```diff
@@ -32,8 +32,12 @@
 }
+
+.selected_puppetclass.unavailable a {
+  color: #d1d1d1;
```
is this a patternfly color? maybe add its name / reference to color pallet.

dLobatog replied:
https://github.com/theforeman/foreman/pull/4048#discussion_r91099151 - should be a patternfly color

Rohoover replied:
pf-black-300 =  #d1d1d1

ohadlevy replied:
can't we simply access it via js/css directly? would make it both readable and if pf colors change we will have less of an headache :)

ohadlevy commented on app/models/puppetclass.rb:
```diff
@@ -41,7 +41,10 @@ class Puppetclass < ActiveRecord::Base
-
+  scope :with_possible_environments, lambda {
+    joins('LEFT OUTER JOIN environment_classes ON puppetclasses.id = environment_classes.puppetclass_id
+           LEFT OUTER JOIN environments ON environment_classes.environment_id = environments.id')
```
why left outer join? wont you get duplicates this way? maybe I'm missing something

bastilian commented:
@Rohoover I've changed this to have the warning now next to the environment select and is updated when it changes:

![screen shot 2016-12-22 at 10 53 52](https://cloud.githubusercontent.com/assets/7757/21421795/06c09cde-c835-11e6-9930-efa6a8acd283.png)

It didn't feel right to have a notification every time this state is detected, I think warning this way is much more effective.

In the Puppet Classes tab the unavailable ones are now only grey out (using the patternfly `$grey-lighter` color):

![screen shot 2016-12-22 at 10 20 34](https://cloud.githubusercontent.com/assets/7757/21420873/53c687a0-c830-11e6-8020-a13329a20ffd.png)

If we have more cases like this, where things are unavailable/inactive, it might be a good idea to have this in `colors.scss` (or `status-colors.scss`).

@ohadlevy you were right, the outer join was not the right thing to use here. The `individual_puppetclasses` method now uses Rails' includes, as it should have in the first place and the scope with join is removed.

dLobatog reviewed (changes requested):
@bastilian This looks basically ready except for the JS function missing unit tests. I haven't tested it much yet, but please move the function to `webpack/`, write a couple of tests (for unavailable_classes.size greater than 0 and equal to 0) and ping me so we can get this merged soon.

Thanks!

bastilian commented:
@dLobatog & @ohadlevy I was reluctant before, because I thought it would be beyond the scope of the PR, if moved to webpack and written as a component. I'm gonna move the function and from now on only add JS under `webpack`. :)

Rohoover commented:
@bastilian My only concern is that the two pieces of messaging below it for Puppet Master and Puppet CA are persistent. The warning for environment may be less identifiable because of the persistent messaging. A possible solution would be to place the persistent messaging in inline help:
http://www.patternfly.org/pattern-library/forms-and-controls/field-level-help/#_

As a note: I'm gong to be addressing with the patternfly team if that icon is correct... there are currently two info icons, I'd like to consolidate them.

bastilian commented:
@Rohoover To be consistent with other forms, where it is next to the label as an (i)-icon tooltip, it does make sense to use the field help for Puppet Master and Puppet CA.

Let me know once you heard back from the patternfly team.

dLobatog commented on app/assets/javascripts/application.js:
```diff
@@ -359,6 +359,7 @@ function update_puppetclasses(element) {
       $('#puppet_klasses').html(request);
       reload_puppetclass_params();
       tfm.tools.activateTooltips();
+      tfm.hostgroups.check_for_unavailable_puppetclasses();
```
`checkForUnavailablePuppetclasses` typo

dLobatog commented on app/assets/stylesheets/puppetclasses.scss:
```diff
@@ -32,8 +32,12 @@
 }
+
+.selected_puppetclass.unavailable a {
+  color: $gray-lighter;
```
Use something from http://www.patternfly.org/styles/color-palette/#_ , the current one is barely visible

`pf-black-400 #bbbbbb` looks nice IMO for disabled elements

dLobatog reviewed (changes requested):
@bastilian I've been looking at it and the functionality is good, it detects the envs just fine (after fixing the typo in application.js). My only comments are:

* The color is barely distinguishable from white, it might be my background, but I think using some color from the patternfly pallette would make it more visible, see the comment inline.
* The warning isn't really noticeable on the "Host" tab, plus as @Rohoover it's confusing to have both persistent and ephemeral information there. Personally I didn't even notice the information when I was testing it, and it wasn't until I came back to the PR that I saw it. What I meant by adding `pf-warning` is that you can put it in the "Puppet classes" tab, on top of everything as @ohadlevy mentioned. That way it's immediately obvious why these are grey, while the current way forces you to go back. 
* It doesn't seem to work in the Host form, which is the same.
* The warning doesn't look good when Katello is installed, see the screenshot.

Long comment, but the fixes needed are quite minimal, thanks @bastilian

dLobatog commented:
The warning with Katello installed:
![screenshot from 2017-01-02 11-29-26](https://cloud.githubusercontent.com/assets/598891/21587744/d9345a68-d0df-11e6-8596-cb4d77bb93c2.png)

bastilian commented:
@dLobatog I've addressed the minor changes and move the warning to the Puppet Classes tab now:

![screen shot 2017-01-02 at 13 11 05](https://cloud.githubusercontent.com/assets/7757/21589100/9c3459c4-d0ee-11e6-8813-0653588f3cc2.png)

One thing that would be good though is to somehow let the user know that "something happened" in the Puppet Classes tab when the environment was changed or do you think it is enough for them to see it only in the tab?

Rohoover commented:
@bastilian 

The correct icon should be pficon_info, not the blue info icon. (For the field level help). 

Also - I think currently the tab turns red when there is an error, not sure if the same holds true for a warning. That said, being that the content has changed in the tab, it should be made known to the user. Though this is not currently defined, it may be good to have something more or other than just a color change - as I have talked to a user who is color-blind, and color only indicators were useless for him.

bastilian commented:
@Rohoover How about adding a warning sign to the tab, instead of it turning red? Like this:

![screen shot 2017-01-04 at 08 03 32](https://cloud.githubusercontent.com/assets/7757/21634045/7fb39730-d254-11e6-8b44-1230af6679ae.png)

Rohoover commented:
@bastilian +1

bastilian commented:
@Rohoover It's now implemented and the screens look as follows when an environment is selected with classes unavailable:

![screen shot 2017-01-04 at 19 48 36](https://cloud.githubusercontent.com/assets/7757/21656119/09a5f5b0-d2bd-11e6-9b07-2b01e6df8b4e.png)
![screen shot 2017-01-04 at 19 48 43](https://cloud.githubusercontent.com/assets/7757/21656120/09aa7338-d2bd-11e6-8b03-bc089f67be5a.png)

dLobatog commented:
@bastilian Awesome - if tests pass, I'll merge (and rebase first, as I solved a small conflict to let tests run).

Great job at figuring out how to show warnings like this with @Rohoover , to be fair we basically relied on making the fields red any time anything wrong happens. As you said that's not going to work for color-blind users. Also, a warning when changing something in the form doesn't have to look the same as an error when you submit it. 

Is there any recommendation on the Patternfly guidelines for forms, or can we make this (warning icon on the tab) a Patternfly recommendation?

dLobatog commented on webpack/assets/javascripts/foreman_hostgroups.js:
```diff
@@ -0,0 +1,21 @@
+    tab.prepend('<span class="pficon pficon-warning-triangle-o"></span> ');
+    puppetKlassesTab.prepend(warning);
+  } else {
+    puppetKlassesTab.find('#puppetclasses_unavaliable_warning').remove();
```
typo (here and above), `unavailable`

dLobatog commented on webpack/assets/javascripts/foreman_hostgroups.js:
```diff
@@ -0,0 +1,21 @@
+      ${warningMessage}
+    </span>`;
+
+  if (unavailableClasses.size() > 0) {
```
Seems like `unavailableClasses.size() > 0` is true even after you change back from the environment that has the classes, so every time I change environments, a new pficon warning is added

dLobatog reviewed (changes requested):
@bastilian Found a problem when switching back to the original environment (where the classes are available), it added another warning icon. Every time I change environments now it adds a new icon basically. I'll try to figure out what's going on

dLobatog commented on app/views/puppetclasses/_selectedClasses.html.erb:
```diff
@@ -1,4 +1,4 @@
-<%= content_tag_for :li, klass, :selected, :class => cycle('even', 'odd') do %>
+<%= content_tag_for :li, klass, :selected, :class => [cycle('even', 'odd'), ('unavailable' unless klass.environments.include?(@environment))] do %>
```
Found the issue. `@environment` is only available on the *first* load of the page. On subsequent loads, it doesn't exist as the host controller didn't set it.

I fixed it locally by adding two lines to `hostgroup_or_environment_selected`, so it's the following:

```ruby

  def hostgroup_or_environment_selected
    Taxonomy.as_taxonomy @organization, @location do
      if params['host']['environment_id'].present? || params['host']['hostgroup_id'].present?
        @environment = Environment.find(params['host']['environment_id'])
        @hostgroup = Hostgroup.find(params['host']['hostgroup_id'])
        render :partial => 'puppetclasses/class_selection', :locals => {:obj => (refresh_host)}
      else
        logger.info "environment_id or hostgroup_id is required to render puppetclasses"
      end
    end
  end
```

I think that's all that's required for it to work.

dLobatog reviewed (changes requested):
See https://github.com/theforeman/foreman/pull/4048/files#r95572403 for the change needed for this to work again

bastilian committed:
Fixes #3551 - Show unavailable Puppet Classes in Host Group edit

Puppet Classes can be assigned to a Host Group even if they are
not available in the set environment.

The Classes persist through out changes of the environment,
but were previously hidden from the UI this changes that.

Unavailable Puppet Classes are now shown in the UI as greyed out
and a warning is displayed to inform the user about them.

dLobatog commented:
Thanks @bastilian @ohadlevy  & @Rohoover !
